### ISSUE TITLE
Buildpack leading us to exceed maximal slug size

### ISSUE THREAD

**Issue by magnusvk** [DataDog/heroku-buildpack-datadog#269]
We are currently dealing with an issue where we cannot deploy our application because our Heroku slug size is too large. Upon investigation it turns out that the Datadog-agent is the biggest contributor to this problem. I am aware of #89 but I am opening a new issue since this is very much not resolved. You say "oh compressed it's only 121 MB"—but that's absurd! That's 40% of the soft limit and 24% of the hard limit, for the datadog agent alone? You *need* to make this more lightweight or else frankly we will need to consider another monitoring service. It is quite disconcerting to see that this is not at all being taken seriously from your side.

**Comment by arapulido** [comment]
Hello @magnusvk! I am really sorry for your experience here. We are working to make this better. I will keep this one open to publish updates as we have them.

**Comment by scottkacyn** [comment]
+1

We're considering switching to Datadog but we're unable to deploy using the Heroku buildpack. We have a monorepo with Ruby (Rails) and React (Node) buildpacks, so including the Datadog agent pushes us over the hard limit. Until this is addressed, we are unable to use your service.

**Comment by arapulido** [comment]
@scottkacyn @magnusvk in parallel to the work we are doing, can I ask you what other buildpacks are you using? If you are using buildpacks that install extra packages (like `heroku-buildpack-apt` for example), those need to be before the Datadog one, if not the Datadog buildpack will take more space than needed. You can read about order of buildpacks here: https://docs.datadoghq.com/agent/basic_agent_usage/heroku/#order-of-buildpacks 

Not sure if this is your case, but this has solved the size problem in the past to other users.

**Comment by scottkacyn** [comment]
@arapulido we use the following buildpacks (in order):

- heroku/nodejs
- heroku/ruby

**Comment by magnusvk** [comment]
```
=== tremdendous Buildpack URLs
1. heroku/ruby
2. https://github.com/DataDog/heroku-buildpack-datadog.git
3. heroku/metrics
```

Should we move up `heroku/metrics`? My guess is it wouldn't make a difference.

**Comment by mobilutz** [comment]
@magnusvk 
Jumping in here with another idea as we just recently reached the hard limit of the slugs and then reduced the size almost by 200MB with these changes:
* add `/spec` to `.slugignore` -> ONLY possible when you don't use HerokuCI!!!
* don't create JS source-maps in production if applicable
* remove `node_modules` folders before slug generation:
https://blog.saeloun.com/2020/05/04/how-to-reduce-heroku-slug-size.html#node_modules-and-other-artifacts
https://github.com/heroku/heroku-buildpack-ruby/issues/792
* remove old compiled webpacker assets:
https://github.com/rails/webpacker/pull/1744
* run build cache cleanup:
  * https://help.heroku.com/18PI5RSY/how-do-i-clear-the-build-cache
  * `heroku repo:purge_cache`
  https://github.com/heroku/heroku-repo#purge-cache

With all of these measures we have a big rails app with one rails engine and some very big react applications running with 282MB
```
-----> Compressing...
       Done: 282.7M
```

**Comment by arapulido** [comment]
@mobilutz Wow! Thanks a lot for those suggestions!

**Comment by gaggle** [comment]
I have a very small API that's already breaking the 500M limit, significantly caused by the DD build pack. These are the build packs in use:

* https://github.com/heroku/heroku-buildpack-google-chrome
* https://github.com/jontewks/puppeteer-heroku-buildpack
* https://github.com/DataDog/heroku-buildpack-datadog
* heroku/nodejs

I've added a ``heroku-postbuild` script to log some size info during the build-process:
```
PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true
DD_AGENT_MAJOR_VERSION=6
DD_PYTHON_VERSION=2
       --- ROOT
       1.3G	.
       965M	./.apt
       175M	./node_modules
       116M	./.heroku
       23M	./src
       ...

       --- .APT
       965M	.apt
       670M	.apt/opt
       398M	.apt/opt/datadog-agent
       274M	.apt/usr
       272M	.apt/opt/google
       ...
```
Here we see the DD agent and Google Chrome taking up most of the space. I've already squeezed Chrome by downloading it via the Google Chrome buildpack because it fetches the headless binary which is smaller than the normal Chrome binary.

Just for good measure here's the output for Agent v7 + Python 3:
```
PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true
DD_AGENT_MAJOR_VERSION=7
DD_PYTHON_VERSION=3
       --- ROOT
       1.3G	.
       975M	./.apt
       175M	./node_modules
       116M	./.heroku
       23M	./src
       ...

       --- .APT
       975M	.apt
       680M	.apt/opt
       408M	.apt/opt/datadog-agent
       274M	.apt/usr
       272M	.apt/opt/google
       201M	.apt/usr/lib
       ...
```
The slight increase in size is unwelcomed but in practical terms they're both very big packs.


These numbers are uncompressed, but the end result is still a large slugsize that Heroku won't deploy. I found it nice how the Puppeteer buildpack allows us to use an externally provided Google Chrome, because that way we can slot in a smaller binary. Perhaps a similar case can be made for the DD buildpack.

**Comment by gaggle** [comment]
In a tragic-lol-sob I'm now exploiting the small size-difference between Agent 7 / Py3 v Agent 6 / Py2. By switching to A6/Py2 I save _just_ enough that my slugsize drops to 499.2M.

So, technically my service runs with full APM tracing now, but at the tiniest error-margin! 😱

It's possible my efforts are better spend finding ways to get off Heroku but if someone has space-saving techniques to suggest I'm happy to try them.

Just to add some more data, here's a full overview of what takes up space when logging into the dyno:
```
~ $ pwd
/app
~ $ du -h .apt | sort -hr | head -n 40
1.3G    .apt
675M    .apt/opt
560M    .apt/usr
487M    .apt/usr/lib
485M    .apt/usr/lib/x86_64-linux-gnu
403M    .apt/opt/datadog-agent
323M    .apt/usr/lib/x86_64-linux-gnu/dri
298M    .apt/opt/datadog-agent/embedded
272M    .apt/opt/google/chrome
272M    .apt/opt/google
229M    .apt/opt/datadog-agent/embedded/lib
213M    .apt/opt/datadog-agent/embedded/lib/python2.7
179M    .apt/opt/datadog-agent/embedded/lib/python2.7/site-packages
105M    .apt/opt/datadog-agent/bin/agent
105M    .apt/opt/datadog-agent/bin
68M     .apt/usr/share
58M     .apt/opt/datadog-agent/embedded/lib/python2.7/site-packages/botocore
57M     .apt/opt/datadog-agent/embedded/lib/python2.7/site-packages/botocore/data
46M     .apt/opt/datadog-agent/embedded/bin
40M     .apt/usr/share/icons
23M     .apt/opt/google/chrome/locales
17M     .apt/usr/share/icons/Humanity
17M     .apt/lib
14M     .apt/opt/datadog-agent/embedded/lib/python2.7/config
13M     .apt/opt/datadog-agent/embedded/lib/python2.7/site-packages/datadog_checks
10M     .apt/opt/google/chrome/WidevineCdm/_platform_specific/linux_x64
10M     .apt/opt/google/chrome/WidevineCdm/_platform_specific
10M     .apt/opt/google/chrome/WidevineCdm
```

From this I note `botocore` looks like a sizable dependency so I hope its inclusion is vitally important.

(FWIW I'm running all this on a Heroku 20 stack)

**Comment by kapilkale** [comment]
Bumping this again. We really like datadog but this is having a negative impact on our ability to test, build, deploy, etc.

@arapulido any idea when someone will even start working on this?

**Comment by luzfcb** [comment]
This issue really blocks the company I work for from using it:

I have some doubts. To use Datadog APM do I really need to configure and use this buildpack?
If so, is there any way to configure it to install only the minimum necessary to run Datadog APM?

**Comment by jezhou** [comment]
Not sure if this helps anyone else, but we use a nodejs/yarn stack at work, and we ran into this issue this week. We fixed it by hooking into the [`heroku-cleanup`](https://devcenter.heroku.com/articles/nodejs-support#heroku-specific-build-steps) step and cleaning up all of the unneeded cached dependencies that were unnecessarily getting compressed into the slug.

For more context: depending on your `yarn` configuration, it will create a `.yarn/cache` folder when you install your dependencies (different from the `node_modules`), which you don't really need when Heroku is finally ready to compress the slug. So what we did was hook into that step and call `yarn cache clean`, which dropped the slug size by a ton. It was more than enough to fit the Datadog agent.

This could be a reasonable workaround for other stacks as well, depending on what artifacts you can safely clean up before compressing the slug. FWIW, I do think this buildpack is still too big (I definitely agree with the top comment).

**Comment by jeremy-lq** [comment]
Hello all. As an update, we're working on a few related changes that will be shipped as part of Agent 7.34.

**Comment by taylordaughtry** [comment]
@jeremy-lq That's great news! Is there a timeline just yet for that release?

**Comment by jeremy-lq** [comment]
@taylordaughtry 7.34 is currently a [release candidate](https://github.com/DataDog/datadog-agent/tags), so final release should be soon.

**Comment by spyrbri** [comment]
https://github.com/DataDog/datadog-agent/releases/tag/7.34.0 Seems like it is released.

**Comment by arapulido** [comment]
The official release will be released during the day. You can watch https://github.com/DataDog/datadog-agent/releases for the official release. Once that happens, the heroku buildpack will follow soon

**Comment by knightq** [comment]
Finally they released it!
Just tried it, still **too heavy** (155MB) to be used by us (it should be at least half the size):

```bash
remote: -----> Downloading Datadog Agent 7.34.0-1        
remote:        Reading package lists...        
remote:        Building dependency tree...        
remote:        Recommended packages:        
remote:          datadog-signing-keys        
remote:        The following NEW packages will be installed:        
remote:          datadog-heroku-agent        
remote:        0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.        
remote:        Need to get 200 MB of archives.        
remote:        After this operation, 655 MB of additional disk space will be used. 
```
...
```bash
remote:  !     Compiled slug size: 587.9M is too large (max is 500M).        
remote:  !     See: http://devcenter.heroku.com/articles/slug-size        
remote: 
remote:  !     Push failed    
```

**Comment by knightq** [comment]
Hurray! It worked 🎉 
Thank you @arapulido and Datadog team!

**Comment by arapulido** [comment]
@knightq  That's great to hear!!!

**Comment by arapulido** [comment]
Hello all!

As we have now the specific Heroku build for the agent and we have released buildpack `2.0` a month ago with it, I will close this issue now.

Feel free to open other specific ones once you test with the new version.

Thanks!
Ara

**Comment by arapulido** [comment]
Hello @ostenbom ! Thanks for your comment. The 90MB refers to the size of the slug once compressed. You can use an empty application adding the buildpack to see how much compressed space it takes.

**Comment by arapulido** [comment]
Hello @ostenbom ! We discovered that the Datadog agent 7.37 introduced new binaries that are not needed by the buildpack and that increased the size. We just released a fix for that. Can you test the latest version of the buildpack and see if it makes the size a bit better? Thanks!
